lib/ScriptRunner.js

```diff
@@ -0,0 +1,260 @@
+            _this._control.sendCommand(cmd);
+        } catch (err) {
+            reject(err);
+        }
+    });
+};
+
+ScriptRunner.prototype.finish = function () {
+    var _this = this;
+    return new Promise(function (fulfill, reject) {
+        try {
+            var noop = function () {
+            };
+            _this._control.on("FINISHED", function () {
+                fulfill();
+            });
+            _this.start().then(noop, noop);
+        } catch (err) {
+            reject(err);
+        }
+    });
+
+};
+
+ScriptRunner.prototype.on = function (state, listener) {
+    // TODO, validate state is a valid name?
+    var arrayOfListeners = this._stateListenerMap[state];
+    if (!arrayOfListeners) {
+        arrayOfListeners = [];
+    }
+    arrayOfListeners.push(listener);
+    this._stateListenerMap[state] = arrayOfListeners;
+};
+
+/**
+ * Internal call to change state, throws exception if can not
+ * change from one valid state to another.
+ * @param newState
+ * @private
+ */
+ScriptRunner.prototype._changeState = function (newState) {
+    var currentState = this._state;
+    // console.trace("Changing from state " + currentState + " to " + newState);
+    switch (newState) {
+        case "INITIAL":
+            throw "Cannot change into INITIAL state";
+        case "PREPARE":
+            if (currentState !== "INITIAL") {
+                throw ("Cannot change into " + newState + " from " + currentState);
+            }
+            break;
+        case "PREPARED":
+            if (currentState !== "PREPARE") {
+                throw ("Cannot change into " + newState + " from " + currentState);
+            }
+            break;
+        case "START":
+            if (currentState !== "PREPARED") {
+                throw ("Cannot change into " + newState + " from " + currentState);
+            }
+            break;
+        case "STARTED":
+            if (currentState !== "START") {
+                throw ("Cannot change into " + newState + " from " + currentState);
+            }
+            break;
+        case "FINISHED":
+            if (currentState !== "PREPARED" && currentState !== "STARTED" && currentState !== "ABORT") {
+                throw ("Cannot change into " + newState + " from " + currentState);
+            }
+            break;
+        case "ABORT":
+            if (currentState === "FINISHED" || currentState === "DISPOSE" || currentState === "DISPOSED") {
+                throw ("Trying to abort while already finished and cleaning up, current state is " + currentState);
+            }
+            break;
+        case "DISPOSE":
+            if (currentState !== "FINISHED" && currentState !== "ERROR") {
+                throw ("Cannot change into " + newState + " from " + currentState);
+            }
+            break;
+        case "DISPOSED":
+            if (currentState !== "DISPOSE") {
+                throw ("Cannot change into " + newState + " from " + currentState);
```

Comment:
This is a generic comment about `throw message;`. JS purists recommend `throw new Error(message);` as it is supported in all the browsers. Here is the relevant text from [the art of throwing Javascript errors](https://www.nczonline.net/blog/2009/03/10/the-art-of-throwing-javascript-errors-part-2/) site:

```
Doing so will cause an error to be thrown, but not all browsers respond the way you’d expect. Firefox, Opera, and Chrome each display an “uncaught exception” message and then include the message string. Safari and Internet Explorer simply throw an “uncaught exception” error and don’t provide the message string at all. Clearly, this is suboptimal from a debugging point of view.
```